File: ndfs-ts/src/endian.ts
/**
 * Big-endian read/write helpers for NDFS binary data.
 *
 * SPDX-License-Identifier: MIT
 * Copyright (c) 1985-2026 Ronny Hansen
 * HackerCorp Labs - https://github.com/HackerCorpLabs
 */

/** Read a 16-bit unsigned integer in big-endian byte order. */
export function readUint16BE(data: Uint8Array, offset: number): number {
  return (data[offset] << 8) | data[offset + 1];
}

/** Read a 32-bit unsigned integer in big-endian byte order. */
export function readUint32BE(data: Uint8Array, offset: number): number {
  return (
    ((data[offset] << 24) |
      (data[offset + 1] << 16) |
      (data[offset + 2] << 8) |
      data[offset + 3]) >>>
    0
  );
}

/** Write a 16-bit unsigned integer in big-endian byte order. */
export function writeUint16BE(
  data: Uint8Array,
  offset: number,
  value: number,
): void {
  data[offset] = (value >>> 8) & 0xff;
  data[offset + 1] = value & 0xff;
}

/** Write a 32-bit unsigned integer in big-endian byte order. */
export function writeUint32BE(
  data: Uint8Array,
  offset: number,
  value: number,
): void {
  data[offset] = (value >>> 24) & 0xff;
  data[offset + 1] = (value >>> 16) & 0xff;
  data[offset + 2] = (value >>> 8) & 0xff;
  data[offset + 3] = value & 0xff;
}

File: ndfs-py/src/ndfs/endian.py
"""
Big-endian read/write helpers for NDFS binary data.

SPDX-License-Identifier: MIT
Copyright (c) 1985-2026 Ronny Hansen
HackerCorp Labs - https://github.com/HackerCorpLabs
"""
from __future__ import annotations

import struct
from typing import Union

_BufType = Union[bytes, bytearray, memoryview]


def read_uint16_be(data: _BufType, offset: int) -> int:
    """Read a 16-bit unsigned integer in big-endian byte order."""
    return struct.unpack_from(">H", data, offset)[0]


def read_uint32_be(data: _BufType, offset: int) -> int:
    """Read a 32-bit unsigned integer in big-endian byte order."""
    return struct.unpack_from(">I", data, offset)[0]


def write_uint16_be(data: bytearray, offset: int, value: int) -> None:
    """Write a 16-bit unsigned integer in big-endian byte order."""
    struct.pack_into(">H", data, offset, value & 0xFFFF)


def write_uint32_be(data: bytearray, offset: int, value: int) -> None:
    """Write a 32-bit unsigned integer in big-endian byte order."""
    struct.pack_into(">I", data, offset, value & 0xFFFFFFFF)
